src/jdk.incubator.code/share/classes/jdk/incubator/code/type/JavaTypeFactory.java

```diff
@@ -0,0 +1,244 @@
+package jdk.incubator.code.type;
+
+import jdk.incubator.code.TypeElement.ExternalizedTypeElement;
+import jdk.incubator.code.type.WildcardType.BoundKind;
+
+import java.lang.constant.ClassDesc;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+
+class JavaTypeFactory implements TypeElementFactory {
+    // Returns JavaType or JavaRef
+    @Override
+    public JavaType constructType(ExternalizedTypeElement tree) {
+        return constructType(null, tree);
+    }
+
+    //    JavaType:
+    //        ClassType                                             // class type
+    //        PrimitiveType                                         // primitive type
+    //        TypeVar                                               // type variable
+    //        '[' JavaType ']'                                      // array type
+    //
+    //    ClassType:
+    //        ClassTypeNoPackage
+    //        Package '.' ClassTypeNoPackage
+    //
+    //    Package:
+    //        ident
+    //        Package '.' ident
+    //
+    //    ClassTypeNoPackage:
+    //        ident                                                 // simple class type
+    //        ident '<' TypeArg* '>'                                // parameterized class type
+    //        ClassType '::' ClassType                               // nested class type
+    //
+    //    PrimitiveType:
+    //        'boolean'
+    //        'char'
+    //        'byte'
+    //        'short'
+    //        'int'
+    //        'float'
+    //        'long'
+    //        'double'
+    //        'void'
+    //
+    //    TypeVar:
+    //        '(' JavaRef ')' TypeVarRest                           // method/constructor type variable
+    //        ClassType TypeVarRest                                 // class type variable
+    //
+    //    TypeVarRest:
+    //        '::' '<' ident '>'
+    //        '::' '<' ident 'extends' JavaType '>'
+    //
+    //    TypeArg:
+    //        '?'                                                   // bivariant type argument
+    //        '?' 'extends' JavaType                                // covariant type argument
+    //        '?' 'super' JavaType                                  // contravariant type argument
+    //        JavaType
+    JavaType constructType(JavaType qualifier, ExternalizedTypeElement tree) {
```

Comment:
As can be seen, the code here is quite complex. This is due to the fact that there's no general 1-1 mapping between type operators and `JavaType` forms. This means that when we parse an `ExternalizedTypeElement` back into a `JavaType` we have to look hard at its structure, to understand what it is, which leads to the spaghetti code in this class.